input/azureeventhub/inputazureeventhub.go

```diff
@@ -0,0 +1,170 @@
+
+	"github.com/tsaikd/gogstash/config"
+	"github.com/tsaikd/gogstash/config/goglog"
+	"github.com/tsaikd/gogstash/config/logevent"
+)
+
+// ModuleName is the name used in config file
+const ModuleName = "azureeventhub"
+
+// InputConfig holds the configuration json fields and internal objects
+type InputConfig struct {
+	config.InputConfig
+	EventHubNamespaceConnectionString string                 `json:"eventhub_namespace_connection_string"`
+	EventHub                          string                 `json:"eventhub"`
+	StorageConnectionString           string                 `json:"storage_connection_string"`
+	StorageContainer                  string                 `json:"storage_container"`
+	ConsumerGroup                     string                 `json:"group"`
+	OffsetEarliest                    bool                   `json:"offset_earliest"`
+	Extras                            map[string]interface{} `json:"extras"`
+}
+
+// DefaultInputConfig returns an InputConfig struct with default values
+func DefaultInputConfig() InputConfig {
+	return InputConfig{
+		InputConfig: config.InputConfig{
+			CommonConfig: config.CommonConfig{
+				Type: ModuleName,
+			},
+		},
+		ConsumerGroup: azeventhubs.DefaultConsumerGroup,
+	}
+}
+
+func InitHandler(
+	ctx context.Context,
+	raw config.ConfigRaw,
+	control config.Control,
+) (config.TypeInputConfig, error) {
+	conf := DefaultInputConfig()
+	err := config.ReflectConfig(raw, &conf)
+	if err != nil {
+		return nil, err
+	}
+
+	conf.Codec, err = config.GetCodec(ctx, config.ConfigRaw{"type": "azureeventhubjson"}, "azureeventhubjson")
+	if err != nil {
+		return nil, err
+	}
+
+	return &conf, nil
+}
+
+func (t *InputConfig) Start(ctx context.Context, msgChan chan<- logevent.LogEvent) (err error) {
+
+	containerClient, err := container.NewClientFromConnectionString(t.StorageConnectionString, t.StorageContainer, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	checkpointStore, err := checkpoints.NewBlobStore(containerClient, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	wg := &sync.WaitGroup{}
+
+	go func() {
+		consumerClient, err := azeventhubs.NewConsumerClientFromConnectionString(t.EventHubNamespaceConnectionString, t.EventHub, t.ConsumerGroup, nil)
+		if err != nil {
+			panic(err)
+		}
+		defer consumerClient.Close(ctx)
+
+		var earliest bool = t.OffsetEarliest
+		var latest bool = !t.OffsetEarliest
+
+		processor, err := azeventhubs.NewProcessor(consumerClient, checkpointStore, &azeventhubs.ProcessorOptions{
+			UpdateInterval:        1 * time.Minute,
+			LoadBalancingStrategy: azeventhubs.ProcessorStrategyBalanced,
+			StartPositions: azeventhubs.StartPositions{
+				Default: azeventhubs.StartPosition{
+					Earliest: &earliest,
+					Latest:   &latest,
+				},
+			},
+		})
+		if err != nil {
+			panic(err)
+		}
+
+		go func() {
+			for {
+				partitionClient := processor.NextPartitionClient(ctx)
+				if partitionClient == nil {
+					break
+				}
+
+				go func() {
+					wg.Add(1)
```

Comment:
it's better to `defer wg.Done()` here, and some of the break in the for loop can return earlier